src/main/scala/it/pagopa/interop/agreementprocess/common/readmodel/ReadModelQueries.scala

```diff
@@ -0,0 +1,133 @@
+import org.mongodb.scala.Document
+import org.mongodb.scala.bson.conversions.Bson
+import org.mongodb.scala.model.Aggregates.{`match`, count, lookup, project, sort, unwind, addFields}
+import org.mongodb.scala.model.{Filters, Field}
+import org.mongodb.scala.model.Projections.{computed, fields, include}
+import org.mongodb.scala.model.Sorts.ascending
+import it.pagopa.interop.agreementmanagement.model.persistence.JsonFormats._
+import it.pagopa.interop.catalogmanagement.{model => CatalogManagement}
+
+object ReadModelQueries {
+  def listAgreements(
+    eServicesIds: List[String],
+    consumersIds: List[String],
+    producersIds: List[String],
+    descriptorsIds: List[String],
+    states: List[AgreementState],
+    showOnlyUpgradeable: Boolean,
+    offset: Int,
+    limit: Int
+  )(readModel: ReadModelService)(implicit ec: ExecutionContext): Future[PaginatedResult[PersistentAgreement]] = {
+
+    val query: Bson =
+      listAgreementsFilters(eServicesIds, consumersIds, producersIds, descriptorsIds, states, showOnlyUpgradeable)
+
+    val filterPipeline: Seq[Bson] = Seq(`match`(query), lookup("eservices", "data.eserviceId", "data.id", "eservices"))
+
+    val upgradablePipeline: Seq[Bson] = {
+      if (showOnlyUpgradeable) {
+        Seq(
+          unwind(fieldName = "eservices.data.descriptors"),
+          addFields(
+            Field(
+              "currentDescriptor",
+              Document("""{ $filter: {
+              input: "$eservices.data.descriptors",
+              as: "descr",         
+              cond: {  $eq: ["$$descr.id" , "$data.descriptorId"]}}} }""")
+            )
+          ),
+          unwind(fieldName = "$currentDescriptor"),
+          addFields(
+            Field(
+              "upgradableDescriptor",
+              Document(
+                """{ $filter: {
+              input: "$eservices.data.descriptors",
+              as: "upgradable",         
+              cond: { $and: [
+                      {$gt:[ {"$toInt" : "$$upgradable.version"} , {"$toInt": "$currentDescriptor.version"}]}, 
+                      {$in:["$$upgradable.state", [""" + CatalogManagement.Published + """, """ +
+                  CatalogManagement.Suspended + """]]}
+                  ]}}} }"""
+              )
+            )
+          ),
+          `match`(Filters.exists("upgradableDescriptor.0", true))
+        )
+
+      } else
+        Seq.empty
+    }
+
+    for {
+      agreements <- readModel.aggregate[PersistentAgreement](
+        "agreements",
+        filterPipeline ++ upgradablePipeline ++
+          Seq(
+            project(
+              fields(include("data"), computed("lowerName", Document("""{ "$toLower" : "$eservices.data.name" }""")))
+            ),
+            sort(ascending("lowerName"))
+          ),
+        offset = offset,
+        limit = limit
+      )
+      count      <- readModel.aggregate[TotalCountResult](
+        "agreements",
+        Seq(
```

Comment:
we need to apply the same filters otherwise the count will not reflect actual results
```suggestion
        if(upgradablePipeline.nonEmpty) filterPipeline ++ upgradablePipeline else Nil ++ Seq(
```